module/icp/include/impl/impl.h

```diff
@@ -0,0 +1,145 @@
+#ifndef	_ALG_IMPL_H
+#define	_ALG_IMPL_H
+
+/*
+ * Common code for managing hash implementations.
+ */
+
+#ifdef	__cplusplus
+extern "C" {
+#endif
+
+#include <sys/zfs_context.h>
+
+/* A function that tests whether method will work. */
+typedef boolean_t (*alg_impl_will_work_f)(void);
+
+/* Maximum length of hash implementation name (including NULL). */
+#define	ALG_IMPL_NAME_MAX (16)
+
+/* Hash implementation operations and name. */
+typedef struct alg_impl_ops {
+	void *ctx; /* Algorithm implementation context. */
+
+	alg_impl_will_work_f is_supported;
+
+	/* Algorithm priorty. Higher is faster. */
+	uint32_t priority;
+
+	char name[ALG_IMPL_NAME_MAX];
+} alg_impl_ops_t;
+
+/* Fastest hash implementation. */
+#define	ALG_IMPL_FASTEST (UINT32_MAX)
+
+/* Cycle to next implementation. */
+#define	ALG_IMPL_CYCLE (UINT32_MAX-1)
+
+/*
+ * Algorithm implementation configuration.
+ */
+typedef struct alg_impl_conf {
+	const alg_impl_ops_t **available; /* Available implementations. */
+	size_t available_n; /* Length of available. */
+
+	const alg_impl_ops_t **supported; /* Supported implementations. */
+	size_t supported_n; /* Length of supported. */
+
+    size_t cycle_impl_idx; /* Index in case of "cycle" implementation. */
+
+	/*
+	 * Index into supported or one of ALG_IMPL_CYCLE or ALG_IMPL_FASTEST.
+	 */
+	uint32_t icp_alg_impl;
+
+	/*
+	 * User desired value of icp_alg_impl, before initialization finishes.
+	 * Must be one of ALG_IMPL_FASTEST or ALG_IMPL_CYCLE.
+	 */
+	uint32_t user_sel_impl;
+
+
+	alg_impl_ops_t fastest; /* Fastest implementation. */
+	alg_impl_ops_t *generic; /* Fallback generic implementation. */
+
+	boolean_t initialized; /* Is configuration initilized. */
+} alg_impl_conf_t;
```

Comment:
Is this abstraction useful?

It could also replace the configuration code in `aes_impl.c`

Reply:
Definitely useful, since it already prevents duplicate code for sha512